### PR TITLE
kyverno policies

### PR DESCRIPTION
**Pull request by Nivedita-coder**
Addresses the following issues: openebs/openebs#3385
Signed-off-by: Nivedita Prasad niveditaprasad81@gmail.com

Steps for installation & testing.
1.Install Kyverno using [YAMLs](https://kyverno.io/docs/installation/#install-kyverno-using-yamls) or [Helm](https://kyverno.io/docs/installation/#install-kyverno-using-helm) in the cluster.
2. Apply Default [kyverno policies](https://github.com/openebs/openebs/pull/3409/files) for OpenEBS using `kubectl apply -f `
3. After that install OpenEBS `kubectl apply -f https://openebs.github.io/charts/openebs-operator.yaml`
4.Testing steps:
Check the bd `kubectl get bd -n openebs`
And, create cStoor pools using the above blockdevices.

### CODE REVIEW

**Review by RealHarshThakur — COMMENTED** [review]
Is installation of kyverno/docs related to it supposed to be a separate PR?

**Review by prateekpandey14 — COMMENTED** [review]
@Nivedita-coder can we update the PR description with details like how to deploy the kyverno and what steps has been followed to tests these policies.

**Comment by Nivedita-coder** [comment]
> @Nivedita-coder can we update the PR description with details like how to deploy the kyverno and what steps has been followed to tests these policies.

Done!

File: kyverno/allow-capabilities.yaml [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+      validate:
+        message: >-
+          Default set of capabilities are allowed.        
+        pattern:
+          spec:
+            containers:
+              - =(securityContext):
+                  =(capabilities):
+                    =(add): "*"
+            =(initContainers):
```
Comment (akhilerm) [inline kyverno/allow-capabilities.yaml:31]: Do we have any initContainers created from openebs? If not, do we need this pattern here?

File: kyverno/allow-host-namespaces.yaml [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+          kinds:
+            - Pod
+      validate:
+        message: >-
+          Sharing the host namespaces is allowed. The fields spec.hostNetwork,
+          spec.hostIPC, and spec.hostPID must be set to true.          
+        pattern:
+          spec:
+            =(hostPID): "true"
+            =(hostIPC): "true"
```
Comment (akhilerm) [inline kyverno/allow-host-namespaces.yaml:31]: Any components which use hostIPC? I am not sure of it. cc: @prateekpandey14

File: kyverno/allow-host-ports.yaml [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+        resources:
+          kinds:
+            - Pod
+      validate:
+        message: >-
+          Use of host ports is allowed. The fields spec.containers[*].ports[*].hostPort
+          and spec.initContainers[*].ports[*].hostPort must be empty.          
+        pattern:
+          spec:
+            =(hostPort):
```
Comment (akhilerm) [inline kyverno/allow-host-ports.yaml:28]: q: If we have hostnetwork enabled, do we still need to explicitly mention about hostports in the policy.?

File: kyverno/allow-privilege-escalation.yaml [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+          name: "openebs-*"
+          kinds:
+            - Pod
+      validate:
+        message: >-
+          Privilege escalation is allowed. The fields spec.containers[*].securityContext.allowPrivilegeEscalation, 
+          and spec.initContainers[*].securityContext.allowPrivilegeEscalation must be defined or set to `true`.
+        pattern:
+          spec:
+            =(initContainers):
```
Comment (akhilerm) [inline kyverno/allow-privilege-escalation.yaml:28]: Here also, I am not sure, if we should add initContainers.

Comment (akhilerm) [reply kyverno/allow-privilege-escalation.yaml:28]: Also, do we need privilege escalation in the spec. AFAIK, we are using only privileged containers.

**Review by akhilerm — COMMENTED** [review]
@Nivedita-coder Have a few questions/comments on whether we should use initContainers in the policy.

File: kyverno/require-user-groups.yaml [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+          Changing to root group ID is allowed.
+        pattern:
+           spec:
+             =(securityContext):
+               =(fsGroup): 'RunAsAny'
+
+
+
+          
+
```
Comment (akhilerm) [inline kyverno/require-user-groups.yaml:66]:
```suggestion
```

File: kyverno/allow-host-namespaces.yaml (hunk 2 above)
Comment (prateekpandey14) [reply kyverno/allow-host-namespaces.yaml:31]: I don't think as well, mostly not required to set.  its may be added here following the [standard template](https://kyverno.io/policies/pod-security/baseline/disallow-host-namespaces/disallow-host-namespaces/)

File: kyverno/allow-capabilities.yaml (hunk 1 above)
Comment (prateekpandey14) [reply kyverno/allow-capabilities.yaml:31]: We can decide its a standard policy pattern which includeds the init containers as well in [kyverno docs.](https://kyverno.io/policies/pod-security/)

File: kyverno/allow-privilege-escalation.yaml (hunk 4 above)
Comment (prateekpandey14) [reply kyverno/allow-privilege-escalation.yaml:28]: We have `allowPrivilegeEscalation: true` in our CSI Drivers daemonsets

File: kyverno/allow-selinux.yaml [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+          spec:
+            =(securityContext):
+              =(seLinuxOptions): "on"
+            =(initContainers):
+              - =(securityContext):
+                  =(seLinuxOptions): "on"
+            containers:
+              - =(securityContext):
+                  =(seLinuxOptions): "on"
+
```
Comment (prateekpandey14) [inline kyverno/allow-selinux.yaml:38]: We need to tests seLinuxOptions policy in openshift based clusters once.

**Comment by kmova** [comment]
Merged into openebs/charts repo at https://github.com/openebs/charts/pull/242

**Pull request closed without merging**